File: 102.二叉树的层序遍历.cpp
/*
 * @lc app=leetcode.cn id=102 lang=cpp
 *
 * [102] 二叉树的层序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
  public:
    vector<vector<int>> levelOrder(TreeNode *root) {
        if (!root)
            return {};
        vector<vector<int>> v;
        queue<TreeNode *> q;
        q.push(root);
        while (q.size()) {
            vector<int> vv;
            vector<TreeNode *> vvv;
            while (q.size()) {
                TreeNode *node = q.front();
                q.pop();
                vv.push_back(node->val);
                if (node->left) {
                    vvv.push_back(node->left);
                }
                if (node->right) {
                    vvv.push_back(node->right);
                }
            }
            v.push_back(vv);
            for(auto &i: vvv) {
                q.push(i);
            }
        }
        return v;
    }
};
// @lc code=end

File: 102.二叉树的层序遍历.go
/*
 * @lc app=leetcode.cn id=102 lang=golang
 *
 * [102] 二叉树的层序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	var result = [][]int{}
	var queue []*TreeNode
	var line = []*TreeNode{root}
	for len(line) > 0 {
		var ret []int
		for len(line) > 0 {
			node := line[0]
			ret = append(ret, node.Val)
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
			line = line[1:]
		}
		line = queue
		queue = nil
		result = append(result, ret)
	}
	return result
}
// @lc code=end

